core/src/main/java/org/apache/accumulo/core/spi/balancer/TabletBalancer.java

```diff
@@ -138,4 +133,15 @@ interface BalanceParameters {
    * @return the time, in milliseconds, to wait before re-balancing.
    */
   long balance(BalanceParameters params);
+
+  /**
+   * Ask the balancer if the tablet is hosted by a TabletServer in the defined resource group
+   *
+   * @return true if TabletServer is in the configured resource group, false otherwise
+   * @since 4.0.0
+   */
+  default boolean isHostedInResourceGroup(Configuration conf, TabletServerId currentLocation,
+      Map<String,Set<TabletServerId>> currentTServerGrouping) {
+    return true;
+  }
```

Comment:
This method should be passed a tablet.  Also conf is probably not needed, that is passed into init and can be kept by the plugin if desired.

Instead of asking if something is in the correct resource group, could we instead ask the balancer what a tablets desired resource group is?

```suggestion
  /**
   * @return the name of the tablet server resource group the balancer would like to see this tablet in
   * @since 4.0.0
   */
  default String getDesiredResourceGroup(TabletId tablet) {
    return default constant;
  }
```

Reply:
IIRC you suggested that we perform this check in TabletManagementIterator (TMI). I can't call `balancer.init()`from the iterator because I don't have access to `ServerContext` to construct a `BalancerEnvironmentImpl`. If someone can point me to a way around this, then I should be able to simplify the method API. I don't think `getDesiredResourceGroup` is a good name as one of the earlier suggestions was to *not* fall back to using the `default` group if the specified group did not have any servers.

This method is called from `TabletMetadata.getTabletState` as it's common to both TGW and TMI, and it seems appropriate because it returns `TabletState.ASSIGNED_TO_WRONG_GROUP`.

I'm not sure how to implement your suggestion.

Reply:
> I can't call balancer.init()from the iterator because I don't have access to ServerContext to construct a BalancerEnvironmentImpl

Could add a constructor like the following.  I think this is ok as this is internal Accumulo code and we know the actual type behind the ServiceEnv interface.

```java
public BalancerEnvironmentImpl(ServiceEnvironment env) {
    super(((ServiceEnvironmentImpl) env).getContext());
  }
```

> I don't think getDesiredResourceGroup is a good name as one of the earlier suggestions was to not fall back to using the default group if the specified group did not have any servers.

Thats a good point, maybe `getRequiredResourceGroup` is a better name.

Reply:
I was able to implement your suggestions in a03bfb6. I modified the TabletBalancer api method to just take a TabletId and I renamed it to `getResourceGroup`.